### PR TITLE
BlockNumberOrHash implementation

### PR DESCRIPTION
**Pull request by Sixtysixter**
The new class BlockNumberOrHash has been added.

### CODE REVIEW

File: silkrpc/core/rawdb/chain.cpp [hunk 1]
```diff
@@ -98,6 +98,19 @@ asio::awaitable<intx::uint256> read_total_difficulty(const DatabaseReader& reade
...
+    if (bnoh.is_number()) {
+        co_return co_await read_block_by_number(reader, bnoh.number());
+    } else if (bnoh.is_hash()) {
+        co_return co_await read_block_by_hash(reader, bnoh.hash());
+    } else if (bnoh.is_tag()) {
+        auto block_number = co_await get_latest_block_number(reader);
+        co_return co_await read_block_by_number(reader, block_number);
+    }
+
+    throw std::runtime_error{"invalid block_number_or_hash value"};
```
Comment (canepat) [inline silkrpc/core/rawdb/chain.cpp:111]: Not sure if this case should be handled as exception or by return value containing `std::optional`. In the end of the day it all depends on how this function is supposed to be used and what must be done when `bnoh` is invalid: I guess that an error should be returned, so probably it's fine as it is

File: silkrpc/types/block.hpp [hunk 2]
```diff
@@ -35,6 +37,70 @@ struct Block : public silkworm::BlockWithHash {
...
+        return is_tag() ? *tag_ : "";
+    }
+
+private:
+    void build(std::string const& bnoh);
+    void set_number(std::string const& input, int base);
+
+    std::unique_ptr<std::uint64_t> number_;
+    std::unique_ptr<evmc::bytes32> hash_;
+    std::unique_ptr<std::string> tag_;
```
Comment (canepat) [inline silkrpc/types/block.hpp:99]: I would really consider the usage of `std::variant` here: whenever possible, we should try to use types already provided in the standard library

File: silkrpc/types/block.hpp [hunk 3]
```diff
@@ -35,6 +37,70 @@ struct Block : public silkworm::BlockWithHash {
 
 std::ostream& operator<<(std::ostream& out, const Block& b);
 
+class BlockNumberOrHash {
+public:
+    BlockNumberOrHash() {}
```
Comment (canepat) [inline silkrpc/types/block.hpp:42]: You could remove default constructor and undefined state, that in my understanding are present just for deserialization from `nlohmann::json` objects, by applying a specific technique outlined [here](https://github.com/nlohmann/json#how-can-i-use-get-for-non-default-constructiblenon-copyable-types)

**Review by canepat — CHANGES_REQUESTED** [review]

File: silkrpc/core/rawdb/chain.cpp (hunk 1 above)
Comment (Sixtysixter) [reply silkrpc/core/rawdb/chain.cpp:111]: This solution has been chosen as it seems in line with other `read_*` functions. that return an exception if something went wrong. 
Returning an `std::optional` introduce a little difference...

File: silkrpc/types/block.hpp (hunk 2 above)
Comment (Sixtysixter) [reply silkrpc/types/block.hpp:99]: knowledge is power!! 
Done.

File: silkrpc/types/block.hpp (hunk 3 above)
Comment (Sixtysixter) [reply silkrpc/types/block.hpp]: Done

**Review by canepat — APPROVED** [review]

**Pull request merged**